### DISCUSSION TITLE
Implement `core_number` for MultiGraph

### DISCUSSION THREAD

**Discussion by jsilter** [networkx/networkx#6378] [Ideas]
I think it would be helpful to have MultiGraph support the `core_number` function, which in turn would allow it to support other [Core](https://networkx.org/documentation/stable/reference/algorithms/core.html) functions. There's some ambiguity around how to treat multiple edges, the simplest option is to be consistent with other `MultiGraph` methods and treat the degree as the total number of edges. If I'm not mistaken it would be a fairly small code change. From this:
```
nbrs[u].remove(v)
pos = node_pos[u]
bin_start = bin_boundaries[core[u]]
node_pos[u] = bin_start
node_pos[nodes[bin_start]] = pos
nodes[bin_start], nodes[pos] = nodes[pos], nodes[bin_start]
bin_boundaries[core[u]] += 1
core[u] -= 1
```
    
To this:
```
nbrs[u].remove(v)
pos = node_pos[u]
bin_start = bin_boundaries[core[u]]
node_pos[u] = bin_start
node_pos[nodes[bin_start]] = pos
nodes[bin_start], nodes[pos] = nodes[pos], nodes[bin_start]
# Changes below this line
num_edges = len(G.get_edge_data(u, v))
bin_boundaries[core[u]] += num_edges
core[u] -= num_edges
```

In the [core_number](https://networkx.org/documentation/stable/_modules/networkx/algorithms/core.html#core_number) function.

**Comment by dschult** [comment]
That's one way to define core number for multigraphs. But there are others. Can you either justify that this is the right number to use, or point to the literature somewhere. 

Typically in cases where it isn't clear how to handle the multiedges, it is better to let the user handle it the way that works best for their problem by forcing them to convert to the Graph in a way that makes sense for their application.  Providing a number for multigraphs without convincing justification for that number makes it *more* difficult for people using multigraphs to find what they actually want -- rather than what is easy to compute.

Thoughts?

**Reply by jsilter** [reply]
nx.MultiGraph defines the degree as the number of edges incident on the given vertex, which is the definition found in Diestel [0]. And nx defines `k_core` (which is my motivating use-case) as: A k-core is a maximal subgraph that contains nodes of degree k or more.
To me this seems like the most obvious and only networkx-consistent way of defining the core number. 

The only other ways I can imagine calculating k-core are:

* Treat parallel edges as redundant, only having them contribute 1 to the degree. For this case, a user would create a `Graph` with the appropriate edges and run `core_number`. My change not needed (and in fact counterproductive, as you say). While I'm sure somebody out there will want to do this, this definition is incredibly unintuitive and I think more people would be surprised to find out parallel edges were treated as redundant rather than additive.

* Consider some subset of edges. Obviously the user would need to define which ones, which would need to be done with a subgraph view (or similar). And from there, calculating the core_number would be easy with my proposed change, and very difficult without it.

I think my change would cover the most common use-cases and have the behavior that most people expect.

[0] Diestel, Reinhard (2005), [Graph Theory](http://www.math.uni-hamburg.de/home/diestel/books/graph.theory/) (3rd ed.), Berlin, New York: Springer-Verlag, [ISBN](https://en.wikipedia.org/wiki/ISBN_(identifier)) [978-3-540-26183-4](https://en.wikipedia.org/wiki/Special:BookSources/978-3-540-26183-4).
